ca_tools/scripts/control_node.py

```diff
@@ -0,0 +1,185 @@
+        self.my_pose.theta = yaw
+        self.move()
+        #rospy.loginfo("My goal is %d, my state is %s",self.goal_num,self.state)
+
+    def move(self):
+        if self.state==TURNING:
+            self.turning()
+        elif self.state==STOP_TURNING:
+            self.stop_turning()
+        elif self.state==FORWARD:
+            self.forward()
+        elif self.state==STOP_FORWARD:
+            self.stop_forward()
+
+    
+    def turning(self):
+
+        if (self.reachedAngle()):
+            self.stop()
+            self.state=STOP_TURNING
+            
+        else:
+            aux = Twist()
+            if (abs(self.my_angleGoal-self.my_pose.theta)>3.14):
+                negateFlag=-1.0
+            else:
+                negateFlag=1.0
+
+            if (self.my_angleGoal-self.my_pose.theta)<0:
+                aux.angular.z=negateFlag*-1.0
+            else:    
+                aux.angular.z=1.0
+
+            rospy.loginfo("negateFlag is %f",negateFlag)
+            aux.angular.x=0.0
+            aux.angular.y=0.0
+            self.my_pub.publish(aux)
+
+    def stop_turning(self):
+        self.stop()
+        self.moveForward()
+        self.state=FORWARD
+
+    def stop(self):
+            aux = Twist()
+            aux.angular.x=0.0
+            aux.angular.y=0.0
+            aux.angular.z=0.0
+            aux.linear.x=0.0
+            aux.linear.y=0.0
+            aux.linear.z=0.0
+            self.my_pub.publish(aux)
+    
+    def moveForward(self):
+        aux = Twist()
+        aux.angular.x=0.0
+        aux.angular.y=0.0
+        aux.angular.z=0.0
+        aux.linear.x=0.2
+        aux.linear.y=0.0
+        aux.linear.z=0.0
+        self.my_pub.publish(aux)
+
+    def forward(self):
+        if (self.reachedPosition()):
+            self.stop()
+            self.state=STOP_FORWARD
+            self.goal_num+=1
+            self.goal_num%=4
+        else:
+            self.setGoalAngle()
+            if (not self.reachedAngle()):
+                self.state=STOP_FORWARD
+                self.stop()
+            else:
+                self.moveForward()
+    
+    def stop_forward(self):
+         self.stop()
+         self.setGoalAngle()
+         self.state=TURNING
+
+    def setGoalAngle(self):
+        self.my_angleGoal=math.atan2((self.my_goals[self.goal_num].y-self.my_pose.y),self.my_goals[self.goal_num].x-self.my_pose.x)
+        rospy.loginfo("Difference between angles is %f,",(self.my_angleGoal-self.my_pose.theta))
+    
+    def reachedAngle(self):
+        return ((abs(self.my_angleGoal-self.my_pose.theta))<0.2 or (abs(self.my_angleGoal-self.my_pose.theta))>6.04)
+
+    def reachedPosition(self):
+        reached_y = ((abs(self.my_goals[self.goal_num].y-self.my_pose.y))<0.2)
+        reached_x=(abs(self.my_goals[self.goal_num].x-self.my_pose.x)<0.2)
+        return (reached_x and reached_y)
+
+
+if __name__ == '__main__':
+    node=ctrl_Node()
+    signal.signal(signal.SIGINT, signal_handler)
```

Comment:
Explain this step, please. Very useful BTW.

Reply:
Commented in the code